### ISSUE TITLE
Complex Bool Query

### ISSUE THREAD

**Issue by marslan390** [danpaz/bodybuilder#204]
Could you please add an example of the following query? Thanks

Select * From Index Where CampaignId = 1 and ((type = "B" and Typeid = 1) or ((type = "C" and Typeid = 1)))

**Comment by conradoramalho** [comment]
Try something like this:

```
 bodybuilder()
    .filter('term', 'CampaignId ', 1)
    .filter('bool', b => b
      .orFilter('bool', b => b
        .filter('term', 'type ', 'B')
        .filter('term', 'Typeid ', 1)
       )
      .orFilter('bool', b => b
        .filter('term', 'type', 'C')
        .filter('term', 'Typeid', 1)
      )
    )
    .build()
```

**Comment by danpaz** [comment]
You should be able to do something like:
```
bodybuilder()
  .orFilter('bool', b => {
    return b.filter('genre', genreA)
      .filter('period', periodA)
  })
  .orFilter('bool', b => {
     return b.filter('genre', genreB)
       .filter('period', periodB)
  })
  .orFilter(...)
```

Which will group your search objects into `bool.should[]` clauses. 

If you need to add filters to your query later in your code, you can hold on to the builder object and only call `.build()` after you're done.

**Comment by t7m** [comment]
Thanks for the reply, but how can I do this in a loop, programmatically? That's the big question.
I don't know how many genre/period/composer combinations I will have.
I already have a bodybuilder object.
How can I add these "and groups" to an "or filter" and then put that "or filter" on the existing bodybuilder object?

**Comment by t7m** [comment]
Oh sorry, maybe I see. Let me try it...

**Comment by t7m** [comment]
Got it, will give it a try. Real quick, can you explain this syntax?

```
  .orFilter('bool', b => {
    return b.filter('genre', 'genreB')
      .filter('period', 'periodB')
  }).build();
```

what is the b => stuff?

Also, if I also want to add a composer, is it as simple as...

    return b.filter('genre', 'genreB')
      .filter('period', 'periodB')
.filter( 'composer', 'blah' )

...?

**Comment by conradoramalho** [comment]
B you have the instance of filter and you can use methods `addFilter, andFilter, filter, notFilter, orFilter`,  whitout need to create a new instance of bodybuilder.

And Yes, you can add more condition into your filter.

@danpaz can correct me if I'm saying something wrong.

Use https://bodybuilder.js.org/, it's very good.

**Comment by t7m** [comment]
One last thing... why do you have 
`.filter( 'period', 'blah' )`
rather than
`filter( 'term', 'period', blah' )`

Is there a reason I've been using "term" this whole time? Is it optional?

**Comment by t7m** [comment]
We've been using https://bodybuilder.js.org/, but without knowing exactly how the syntax should be, we were unable to reach these conclusions. My code is building and I should have a good test in a few minutes

**Comment by conradoramalho** [comment]
No, because I copied the @danpaz example, but I suggest you use what you need, ex: term, match ...

In my opinion, you should use kibana to test your query and the bodybuilder to clean your code and  manipulate conditions.

**Comment by t7m** [comment]
My first attempt gave me 
parsing_exception: bool query does not support [query]

I'm not entirely sure if that was on the ...conditions being added to mainQuery or one of the filters. I'll figure it out and let you know when I finally get the answer. Really appreciate the help

**Comment by conradoramalho** [comment]
It's ok. If you need help you can ask me.

**Comment by t7m** [comment]
Unfortunately, I'm not any closer on this. When I put this in the code...

  mainQuery.andFilter( 'bool', { 
   ...categoryFilter.build()
  });

... I get  [bool] query does not support [query]

bodybuilder shows that this "categoryFilter" is going above the rest of the json body, rather than in with the other filters.

**Comment by t7m** [comment]
It seems that no matter what I do, I get parsing_exception errors. The creation of the second bodybuilder seems like it creates too many extra bool.query.bool.filter.whatever things in the json.

**Comment by johannes-scharlach** [comment]
This is the typical syntax:
```js
bodybuilder().orFilter('bool', b => {
    return b.filter('term', 'genre', 'genreB')
      .filter('term', 'period', 'periodB')
  }).build();
```

You can use other variables from the closure to generate the nested query, like:

```js
const genres = ['jazz', 'punk']

bodybuilder().orFilter('bool', b => {
    for (const genre of genres) {
      b.filter('term', 'genre', genre)
    }

    return b
  }).build();
```
Which is what I would usually do in a case like yours.


The following version also works, which is what @conradoramalho suggested for you to define the `nested` stuff outside of the closure, but you need to use the `.query.bool`, otherwise you expect the syntax errors that you're getting.
```js
const nested = bodybuilder()

nested
    .filter('term', 'genre', 'genreB')
    .filter('term', 'period', 'periodB')
    .build()

const result = bodybuilder().orFilter('bool', {
  ...nested.query.bool
}).build();
```

If you need more help, please post how you're using bodybuilder and the query it generates.

**Comment by t7m** [comment]
Or actually not quite... I need all those orFilters inside an andFilter. So I'm still not quite there.

**Comment by t7m** [comment]
Probably wanting this...

```
bodybuilder().andFilter('bool', b => {
    for (const genre of genres) {
      b.orFilter('term', 'genre', genre)
    }

    return b
  }).build();
```

**Comment by johannes-scharlach** [comment]
Yes, that sounds reasonable

**Comment by t7m** [comment]
Hey, THANK YOU to everyone who commented... after helping me with the syntax, I finally figured out what I needed. If it helps anyone else...

So we're building the main "query" object, and if there are any genres to be searched, we do...
```
        query.andFilter( 'bool', b => {
          for ( const fullUid of genres ) {
            let fullUidArr = fullUid.split( '/' );
            let genreUid= fullUidArr[ 0 ], periodUid = null, composerUid = null;
            if ( fullUidArr.length > 1 ) periodUid = fullUidArr[ 1 ];
            if ( fullUidArr.length > 2 ) composerUid = fullUidArr[ 2 ];

            if ( periodUid == null && composerUid == null ) {
              b.orFilter( 'bool', c => {
                c.andFilter( 'term', 'GenreUid', genreUid );

                return c;
              });
            } else if ( composerUid == null ) {
              b.orFilter( 'bool', c => {
                c.andFilter( 'term', 'GenreUid', genreUid )
                  .andFilter( 'term', 'PeriodUid', periodUid );

                return c;
              });
            } else {
              b.orFilter( 'bool', c => {
                c.andFilter( 'term', 'GenreUid', catUid )
                  .andFilter( 'term', 'PeriodUid', periodUid )
                  .andFilter( 'match', 'ComposersUid', composerUid);

                return c;
              });
            }
          }

          return b;
        });
		
```
